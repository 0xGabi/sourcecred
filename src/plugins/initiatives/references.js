// @flow

import type {URL} from "./initiative";
import type {ReferenceDetector} from "./createGraph";
import {NodeAddress, type NodeAddressT} from "../../core/graph";
import {githubOwnerPattern, githubRepoPattern} from "../../core/repoId";
import type {ReadRepository} from "../discourse/mirrorRepository";
import {topicAddress, userAddress, postAddress} from "../discourse/address";
import {linksToReferences} from "../discourse/references";

// TODO: this should probably be in the core, when universal reference detection is added.
export class CascadingReferenceDetector implements ReferenceDetector {
  refs: $ReadOnlyArray<ReferenceDetector>;

  constructor(refs: $ReadOnlyArray<ReferenceDetector>) {
    this.refs = refs;
  }

  addressFromUrl(url: URL): ?NodeAddressT {
    for (const ref of this.refs) {
      const addr = ref.addressFromUrl(url);
      if (addr) {
        return addr;
      }
    }
  }
}

const urlRegex = new RegExp(
  `^https?://github.com/(${githubOwnerPattern})/(${githubRepoPattern})/(issues|pull)/(\\d+)`
);

// TODO: this is a temporary implementation that needs replacing.
export class GithubReferenceDetector implements ReferenceDetector {
  addressFromUrl(url: URL): ?NodeAddressT {
    const mathes = urlRegex.exec(url);
    if (!mathes) {
      return null;
    }

    const [_, owner, repo, type, number] = mathes;
    return NodeAddress.fromParts([
      "sourcecred",
      "github",
      type === "issues" ? "ISSUE" : "PULL",
      owner,
      repo,
      number,
    ]);
  }
}

// TODO: this is a temporary implementation that needs replacing.
export class DiscourseReferenceDetector implements ReferenceDetector {
  data: ReadRepository;

  constructor(data: ReadRepository) {
    this.data = data;
  }

  addressFromUrl(url: URL): ?NodeAddressT {
    let dst: NodeAddressT | null = null;
    const [reference] = linksToReferences([url]);
    if (!reference) {
      return null;
    }

    switch (reference.type) {
      case "TOPIC": {
        dst = topicAddress(reference.serverUrl, reference.topicId);
        break;
      }
      case "POST": {
        const referredPostId = this.data.findPostInTopic(
          reference.topicId,
          reference.postIndex
        );
        if (referredPostId == null) {
          // Maybe a bad link, or the post or topic was deleted.
          return null;
        }
        dst = postAddress(reference.serverUrl, referredPostId);
        break;
      }
      case "USER": {
        dst = userAddress(reference.serverUrl, reference.username);
        break;
      }
      default: {
        throw new Error(
          `Unexpected reference type: ${(reference.type: empty)}`
        );
      }
    }

    return dst;
  }
}
