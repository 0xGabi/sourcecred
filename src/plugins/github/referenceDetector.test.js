// @flow

import Database from "better-sqlite3";
import path from "path";
import {RelationalView} from "./relationalView";
import {
  GithubReferenceDetector,
  fromRelationalView,
  referenceMapFromDB,
} from "./referenceDetector";
import {MappedReferenceDetector} from "../../core/references";
import {exampleRelationalView} from "./example/example";
import {NodeAddress} from "../../core/graph";

function mapValues<A, B, C>(map: Map<A, B>, fn: (B) => C): Map<A, C> {
  const newMap = new Map();
  for (const [k, v] of map.entries()) {
    newMap.set(k, fn(v));
  }
  return newMap;
}

describe("plugins/github/referenceDetector", () => {
  describe("GithubReferenceDetector", () => {
    it("should be a MappedReferenceDetector", () => {
      // Given
      const map = new Map();

      // When
      const refs = new GithubReferenceDetector(map);

      // Then
      expect(refs).toBeInstanceOf(MappedReferenceDetector);
    });
  });

  describe("fromRelationalView", () => {
    it("should use urlReferenceMap to create the instance", () => {
      // Given
      const rv = new RelationalView();
      const urlReferenceMap = jest.spyOn(rv, "urlReferenceMap");

      // When
      const refs = fromRelationalView(rv);

      // Then
      expect(refs).toBeInstanceOf(MappedReferenceDetector);
      expect(urlReferenceMap).toBeCalledTimes(1);
    });
  });

  describe("referenceMapFromDB", () => {
    it("should work with an example DB, compared to the example RelationalView", () => {
      // Given
      const owner = "sourcecred-test";
      const db = new Database(
        path.join(__dirname, "example", "example-mirror.db")
      );
      const rv = exampleRelationalView();

      // When
      const refs = referenceMapFromDB(db, owner);

      // Then
      const rvRefs = rv.urlReferenceMap();
      expect(refs).toEqual(rvRefs);
      const noNullRefs = mapValues(refs, NodeAddress.toParts);
      expect(noNullRefs).toMatchSnapshot();
    });
  });
});
