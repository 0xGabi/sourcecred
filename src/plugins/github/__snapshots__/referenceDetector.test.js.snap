// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugins/github/referenceDetector referenceMapFromDB should work with an example DB, compared to the example RelationalView 1`] = `
Map {
  "https://github.com/decentralion" => Array [
    "sourcecred",
    "github",
    "USERLIKE",
    "USER",
    "decentralion",
  ],
  "https://github.com/wchargin" => Array [
    "sourcecred",
    "github",
    "USERLIKE",
    "USER",
    "wchargin",
  ],
  "https://github.com/credbot" => Array [
    "sourcecred",
    "github",
    "USERLIKE",
    "BOT",
    "credbot",
  ],
  "https://github.com/sourcecred-test/example-github/commit/6bd1b4c0b719c22c688a74863be07a699b7b9b34" => Array [
    "sourcecred",
    "github",
    "COMMIT",
    "MDY6Q29tbWl0MTIzMjU1MDA2OjZiZDFiNGMwYjcxOWMyMmM2ODhhNzQ4NjNiZTA3YTY5OWI3YjliMzQ=",
  ],
  "https://github.com/sourcecred-test/example-github/commit/0a223346b4e6dec0127b1e6aa892c4ee0424b66a" => Array [
    "sourcecred",
    "github",
    "COMMIT",
    "MDY6Q29tbWl0MTIzMjU1MDA2OjBhMjIzMzQ2YjRlNmRlYzAxMjdiMWU2YWE4OTJjNGVlMDQyNGI2NmE=",
  ],
  "https://github.com/sourcecred-test/example-github/commit/6d5b3aa31ebb68a06ceb46bbd6cf49b6ccd6f5e6" => Array [
    "sourcecred",
    "github",
    "COMMIT",
    "MDY6Q29tbWl0MTIzMjU1MDA2OjZkNWIzYWEzMWViYjY4YTA2Y2ViNDZiYmQ2Y2Y0OWI2Y2NkNmY1ZTY=",
  ],
  "https://github.com/sourcecred-test/example-github/commit/c430bd74455105f77215ece51945094ceeee6c86" => Array [
    "sourcecred",
    "github",
    "COMMIT",
    "MDY6Q29tbWl0MTIzMjU1MDA2OmM0MzBiZDc0NDU1MTA1Zjc3MjE1ZWNlNTE5NDUwOTRjZWVlZTZjODY=",
  ],
  "https://github.com/sourcecred-test/example-github/commit/ec91adb718a6045b492303f00d8e8beb957dc780" => Array [
    "sourcecred",
    "github",
    "COMMIT",
    "MDY6Q29tbWl0MTIzMjU1MDA2OmVjOTFhZGI3MThhNjA0NWI0OTIzMDNmMDBkOGU4YmViOTU3ZGM3ODA=",
  ],
  "https://github.com/sourcecred-test/example-github/commit/ecc889dc94cf6da17ae6eab5bb7b7155f577519d" => Array [
    "sourcecred",
    "github",
    "COMMIT",
    "MDY6Q29tbWl0MTIzMjU1MDA2OmVjYzg4OWRjOTRjZjZkYTE3YWU2ZWFiNWJiN2I3MTU1ZjU3NzUxOWQ=",
  ],
  "https://github.com/sourcecred-test/example-github" => Array [
    "sourcecred",
    "github",
    "REPO",
    "sourcecred-test",
    "example-github",
  ],
  "https://github.com/sourcecred-test/example-github/issues/1" => Array [
    "sourcecred",
    "github",
    "ISSUE",
    "sourcecred-test",
    "example-github",
    "1",
  ],
  "https://github.com/sourcecred-test/example-github/issues/2" => Array [
    "sourcecred",
    "github",
    "ISSUE",
    "sourcecred-test",
    "example-github",
    "2",
  ],
  "https://github.com/sourcecred-test/example-github/issues/4" => Array [
    "sourcecred",
    "github",
    "ISSUE",
    "sourcecred-test",
    "example-github",
    "4",
  ],
  "https://github.com/sourcecred-test/example-github/issues/6" => Array [
    "sourcecred",
    "github",
    "ISSUE",
    "sourcecred-test",
    "example-github",
    "6",
  ],
  "https://github.com/sourcecred-test/example-github/issues/7" => Array [
    "sourcecred",
    "github",
    "ISSUE",
    "sourcecred-test",
    "example-github",
    "7",
  ],
  "https://github.com/sourcecred-test/example-github/issues/8" => Array [
    "sourcecred",
    "github",
    "ISSUE",
    "sourcecred-test",
    "example-github",
    "8",
  ],
  "https://github.com/sourcecred-test/example-github/issues/10" => Array [
    "sourcecred",
    "github",
    "ISSUE",
    "sourcecred-test",
    "example-github",
    "10",
  ],
  "https://github.com/sourcecred-test/example-github/issues/11" => Array [
    "sourcecred",
    "github",
    "ISSUE",
    "sourcecred-test",
    "example-github",
    "11",
  ],
  "https://github.com/sourcecred-test/example-github/issues/12" => Array [
    "sourcecred",
    "github",
    "ISSUE",
    "sourcecred-test",
    "example-github",
    "12",
  ],
  "https://github.com/sourcecred-test/example-github/issues/13" => Array [
    "sourcecred",
    "github",
    "ISSUE",
    "sourcecred-test",
    "example-github",
    "13",
  ],
  "https://github.com/sourcecred-test/example-github/pull/3" => Array [
    "sourcecred",
    "github",
    "PULL",
    "sourcecred-test",
    "example-github",
    "3",
  ],
  "https://github.com/sourcecred-test/example-github/pull/5" => Array [
    "sourcecred",
    "github",
    "PULL",
    "sourcecred-test",
    "example-github",
    "5",
  ],
  "https://github.com/sourcecred-test/example-github/pull/9" => Array [
    "sourcecred",
    "github",
    "PULL",
    "sourcecred-test",
    "example-github",
    "9",
  ],
  "https://github.com/sourcecred-test/example-github/pull/5#pullrequestreview-100313899" => Array [
    "sourcecred",
    "github",
    "REVIEW",
    "sourcecred-test",
    "example-github",
    "5",
    "100313899",
  ],
  "https://github.com/sourcecred-test/example-github/pull/5#pullrequestreview-100314038" => Array [
    "sourcecred",
    "github",
    "REVIEW",
    "sourcecred-test",
    "example-github",
    "5",
    "100314038",
  ],
  "https://github.com/sourcecred-test/example-github/issues/2#issuecomment-373768703" => Array [
    "sourcecred",
    "github",
    "COMMENT",
    "ISSUE",
    "sourcecred-test",
    "example-github",
    "2",
    "373768703",
  ],
  "https://github.com/sourcecred-test/example-github/issues/2#issuecomment-373768850" => Array [
    "sourcecred",
    "github",
    "COMMENT",
    "ISSUE",
    "sourcecred-test",
    "example-github",
    "2",
    "373768850",
  ],
  "https://github.com/sourcecred-test/example-github/issues/2#issuecomment-385576185" => Array [
    "sourcecred",
    "github",
    "COMMENT",
    "ISSUE",
    "sourcecred-test",
    "example-github",
    "2",
    "385576185",
  ],
  "https://github.com/sourcecred-test/example-github/issues/2#issuecomment-385576220" => Array [
    "sourcecred",
    "github",
    "COMMENT",
    "ISSUE",
    "sourcecred-test",
    "example-github",
    "2",
    "385576220",
  ],
  "https://github.com/sourcecred-test/example-github/issues/2#issuecomment-385576248" => Array [
    "sourcecred",
    "github",
    "COMMENT",
    "ISSUE",
    "sourcecred-test",
    "example-github",
    "2",
    "385576248",
  ],
  "https://github.com/sourcecred-test/example-github/issues/2#issuecomment-385576273" => Array [
    "sourcecred",
    "github",
    "COMMENT",
    "ISSUE",
    "sourcecred-test",
    "example-github",
    "2",
    "385576273",
  ],
  "https://github.com/sourcecred-test/example-github/issues/2#issuecomment-385576920" => Array [
    "sourcecred",
    "github",
    "COMMENT",
    "ISSUE",
    "sourcecred-test",
    "example-github",
    "2",
    "385576920",
  ],
  "https://github.com/sourcecred-test/example-github/issues/2#issuecomment-385576936" => Array [
    "sourcecred",
    "github",
    "COMMENT",
    "ISSUE",
    "sourcecred-test",
    "example-github",
    "2",
    "385576936",
  ],
  "https://github.com/sourcecred-test/example-github/issues/6#issuecomment-373768442" => Array [
    "sourcecred",
    "github",
    "COMMENT",
    "ISSUE",
    "sourcecred-test",
    "example-github",
    "6",
    "373768442",
  ],
  "https://github.com/sourcecred-test/example-github/issues/6#issuecomment-373768538" => Array [
    "sourcecred",
    "github",
    "COMMENT",
    "ISSUE",
    "sourcecred-test",
    "example-github",
    "6",
    "373768538",
  ],
  "https://github.com/sourcecred-test/example-github/issues/6#issuecomment-385223316" => Array [
    "sourcecred",
    "github",
    "COMMENT",
    "ISSUE",
    "sourcecred-test",
    "example-github",
    "6",
    "385223316",
  ],
  "https://github.com/sourcecred-test/example-github/issues/6#issuecomment-417104047" => Array [
    "sourcecred",
    "github",
    "COMMENT",
    "ISSUE",
    "sourcecred-test",
    "example-github",
    "6",
    "417104047",
  ],
  "https://github.com/sourcecred-test/example-github/issues/11#issuecomment-420811872" => Array [
    "sourcecred",
    "github",
    "COMMENT",
    "ISSUE",
    "sourcecred-test",
    "example-github",
    "11",
    "420811872",
  ],
  "https://github.com/sourcecred-test/example-github/issues/11#issuecomment-420813013" => Array [
    "sourcecred",
    "github",
    "COMMENT",
    "ISSUE",
    "sourcecred-test",
    "example-github",
    "11",
    "420813013",
  ],
  "https://github.com/sourcecred-test/example-github/issues/11#issuecomment-420813206" => Array [
    "sourcecred",
    "github",
    "COMMENT",
    "ISSUE",
    "sourcecred-test",
    "example-github",
    "11",
    "420813206",
  ],
  "https://github.com/sourcecred-test/example-github/issues/11#issuecomment-420813621" => Array [
    "sourcecred",
    "github",
    "COMMENT",
    "ISSUE",
    "sourcecred-test",
    "example-github",
    "11",
    "420813621",
  ],
  "https://github.com/sourcecred-test/example-github/pull/3#issuecomment-369162222" => Array [
    "sourcecred",
    "github",
    "COMMENT",
    "PULL",
    "sourcecred-test",
    "example-github",
    "3",
    "369162222",
  ],
  "https://github.com/sourcecred-test/example-github/pull/5#issuecomment-396430464" => Array [
    "sourcecred",
    "github",
    "COMMENT",
    "PULL",
    "sourcecred-test",
    "example-github",
    "5",
    "396430464",
  ],
  "https://github.com/sourcecred-test/example-github/pull/5#discussion_r171460198" => Array [
    "sourcecred",
    "github",
    "COMMENT",
    "REVIEW",
    "sourcecred-test",
    "example-github",
    "5",
    "100313899",
    "171460198",
  ],
}
`;
